Port `ConstrainedReschedule` to Rust

phalakbh opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Rebased on top of #15276

This PR ports the `ConstrainedReschedule` pass to Rust, preserving the original logic and maintaining identical user-facing behavior.

Since the pass assumes it is receiving its durations in terms of `dt`, it will now print an error back to Python asking to provide the durations in the correct terms.

### Details and comments
Fixes #12268

phalakbh committed:
Initial commit, port constrained reschedule pass to Rust

phalakbh committed:
Complete remaining portions of the logic implementation

phalakbh committed:
Fix lint

phalakbh committed:
Fix lint

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+
+use crate::target::Target;
+use crate::TranspilerError;
+use ::hashbrown::HashSet;
```
A minor style thing you don't need the leading `::` here.
```suggestion
use hashbrown::HashSet;
```

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+            )
+        })
+        .collect()
+}
```
How does this differ from something like `dag.quantum_successors.filter(|idx| matches!(dag[idx]`, NodeType::Operation(_))`? 

There is also an `op_successors` method on `DAGCircuit` but it's for python. If that is what you're looking for I'd suggest adding it as a `DAGCircuit` method for rust only and work with an iterator return so we can avoid the allocation.

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+#[pyo3(name="constrained_reschedule", signature=(dag, node_start_time, clbit_write_latency, acquire_align, pulse_align, target))]
+pub fn run_constrained_reschedule(
+    dag: &DAGCircuit,
+    node_start_time: &Bound<PyDict>,
```
Is there a rust native type we can use for this? I realize this is coming from the property set in Python. But you're porting the scheduling passes to Rust right? I would expect the asap and alap scheduling passes to be returning a rust struct that is a pyclass that implements the mapping protocol so it looks like a dict to python, but internally it's a hashmap or something for rust so we can avoid the conversion and python over when accessing the start time data.

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+    let op = match dag.dag().node_weight(node_index) {
+        Some(NodeType::Operation(op)) => op,
+        _ => panic!("topological_op_nodes() should only return instances of DagOpNode."),
+    };
```
```suggestion
    let NodeType::Operation(op) = dag[node_index] else {
        unreachable!("topological_op_nodes() should only return operations.")
    };
```

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+        OperationRef::StandardInstruction(StandardInstruction::Reset)
+        | OperationRef::StandardInstruction(StandardInstruction::Measure) => Some(acquire_align),
+        OperationRef::StandardInstruction(StandardInstruction::Delay(_)) => None,
+        _ => None,
```
Is there a reason you need both arms here, since `_` would cover delay already and they both return `None`

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+        node_start_time
+            .set_item(node_index.index(), this_t0)
+            .unwrap();
+    }
```
I'm surprised this is all floats, shouldn't this be an integer? I thought all the alignment constraints were in terms of `dt` which should all be in integers

mtreinish commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,304 @@
+                        .unwrap_or(0.0)
+                }
+                Param::Float(f) => *f,
+                _ => 0.0,
```
Is this correct? If it's not a float or an obj then it's a parameter expression which I believe is supported (for better or worse), that doesn't mean the duration is zero does it? Looking at the python code I don't think anything would fail, it wouldn't make any sense and returning an error might make more sense to handle a parameterized delay.

mtreinish reviewed (commented):
Thanks for doing this, it's a good start. I took a quick look through the PR and left a few inline comments. The biggest concern I have isn't specifically about this PR but more the ALAP and ASAP scheduling passes and what we're putting on the property set. As we're moving the scheduling stage to rust we should be storing our output in rust data structures and make Rust the source of truth, this is how we get both better performance and also expose these passes to C in the future. Right now all the data around start times is in a pydict and that is causing a lot of overhead and also a runtime dependency on Python which we want to eliminate. The simplest path out of this is to change the return type from the scheduling passes to be a rust struct that is exposed as a pyclass but wraps a hashmap or something so we can query and override the start times without Python. A maybe simpler/more isolated solution would be to write this struct and have a python method or function that can build it in python from a dict, then pass the constructed object to the rust code so the rust code doesn't have to interact with a python dict directly.

The other thing which stood out was the use of `f64` for all the durations. Realistically for this pass I would expect everything to be a `u64` or `u32` because I would expect all the gate durations for should all be in terms of dt which is an integer.

phalakbh committed:
Implement suggested changes

phalakbh committed:
Fix lint

phalakbh committed:
Fix lint

jakelishman commented:
Ray: can you check if you want to continue with this PR in situ, or do you want to make a new PR to take this in a different direction?

raynelfss commented:
@jakelishman I'll let you know if anything changes but I will probably just continue with it

raynelfss committed:
Fix: Use new `NodeDurations` struct in `ConstrainedReschedule`

raynelfss committed:
Fix: Redundant casting.

raynelfss committed:
Docs: Update header string to use https

alexanderivrii commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,297 @@
+    dag: &DAGCircuit,
+    node_index: NodeIndex,
+    node_start_time: &mut IndexMap<NodeIndex<u32>, u64, RandomState>,
+    clbit_write_latency: u32,
+    pulse_align: u32,
+    acquire_align: u32,
+    target: Option<&Target>,
+) -> PyResult<()> {
+    let NodeType::Operation(op) = &dag[node_index] else {
+        unreachable!("topological_op_nodes() should only return operations.")
+    };
+
+    let op_view = op.op.view();
+    let alignment = match op_view {
+        OperationRef::Gate(_) | OperationRef::StandardGate(_) => Some(pulse_align),
+        OperationRef::StandardInstruction(StandardInstruction::Reset)
+        | OperationRef::StandardInstruction(StandardInstruction::Measure) => Some(acquire_align),
+        _ => None,
+    };
```
This does not handle things like `OperationRef::Unitary` or the new `PauliProductRotation` and `PauliProductMeasurement`. Do you know what happens when such a gate is present -- does the pass ignore these gates or raises an error?

It probably does not matter. Even if the Python version may technically handle UnitaryGates (I have not tried, but these subclass Gate), I believe that in the normal transpilation flow we shouldn't have any unitary gates when we run the pass. And I think the pass makes even less sense in the context of PPR/PPM gates.

raynelfss replied:
I believe this is the way it was done in Python. If the operation wasn't either a Gate or a Measure or Reset. Although from the looks of it it doesn't deal with delay's the same way it does on Python.

https://github.com/Qiskit/qiskit/blob/c90f59d0d5f009c74576b509fab79bda88e14325/qiskit/transpiler/passes/scheduling/alignments/reschedule.py#L124-L133

**Edit:** I see you reference this in your comment below

alexanderivrii commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,297 @@
+        OperationRef::Gate(_) | OperationRef::StandardGate(_) => Some(pulse_align),
+        OperationRef::StandardInstruction(StandardInstruction::Reset)
+        | OperationRef::StandardInstruction(StandardInstruction::Measure) => Some(acquire_align),
+        _ => None,
```
A minor difference here between Python and Rust versions: the Python version raised a transpiler error (when the op is not Delay or directive), the Rust version does not.

raynelfss replied:
From what I could find, it might have come from [a comment Matthew made before](https://github.com/Qiskit/qiskit/pull/14883#discussion_r2274566070)

raynelfss replied:
Addressed in 08839b9ccfecb0c642934aab9876b4c5bc74d668

alexanderivrii commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,297 @@
+    if let Some(alignment) = alignment {
+        let misalignment = this_t0 % alignment as u64;
+        let shift = if misalignment != 0 {
+            alignment as u64 - misalignment
```
The python version here had 
```python
            if misalignment != 0:
                shift = max(0, alignment - misalignment)
```

I don't know if the difference can ever be negative, so I am wondering if `saturating_sub` would be a better fit here.

raynelfss replied:
I think the difference here is that these are unsigned integers, so it **shouldn't** be less than zero. However, I think if the case happened whjere it is less than zero, we'd end with an overflow. So using `saturating_sub` here might be the way to go.

raynelfss replied:
Added in 08839b9ccfecb0c642934aab9876b4c5bc74d668

alexanderivrii commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,297 @@
+            .collect();
+        let duration = target.get_duration(op.op.name(), &qargs).unwrap_or(0.0);
+        this_t0 + duration as u64
+    } else if matches!(
+        op_view,
+        OperationRef::StandardInstruction(StandardInstruction::Delay(_))
+    ) {
+        let params = op.params_view();
+        let param = params
+            .first()
+            .ok_or_else(|| PyValueError::new_err("Delay instruction missing duration parameter"))?;
+        let duration = match param {
+            Param::Obj(val) => {
+                // Try to extract as different numeric types
+                Python::attach(|py| val.bind(py).extract::<u64>())
+            }
+            Param::Float(f) => Ok(*f as u64),
+            _ => Err(TranspilerError::new_err(
+                "The provided Delay duration is not in terms of dt.",
+            )),
+        }?;
+
+        this_t0 + duration
```
This seems a bit too complicating for replacing the following two lines in python code:

```python
        elif node.name == "delay":
            new_t1q = this_t0 + node.op.duration
```

In particular, can `param` be really of type `obj`?

raynelfss replied:
Yeah, this kind of shows that we need a `Param:Int()` variant for `Param` (or just to normalize using `Symbol::Int`).

alexanderivrii commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,297 @@
+
+        // Compute overlap if there is qubits overlap
+        let qreg_overlap = if !this_qubits.is_disjoint(&next_qubits) {
+            new_t1q - next_t0q
```
To double-check: this can never underflow, right? I.e. `new_t1q` is always greater than or equal to `next_t0q`?

raynelfss replied:
It shouldn't overflow as t0 **should** be smaller than t1 since we don't necessarily have negative times recorded in the `node_start_times` property. And we strictly add things onto t1. Do you think it would be worth to add a guard here??

alexanderivrii replied:
I agree that t0 <= t1 should follow from the algorithm, I am fine keeping this as is

alexanderivrii commented on crates/transpiler/src/passes/constrained_reschedule.rs:
```diff
@@ -0,0 +1,297 @@
+            && !this_clbits.is_disjoint(&next_clbits)
+        {
+            if let (Some(t1c), Some(t0c)) = (new_t1c, next_t0c) {
+                t1c - t0c
```
same question on whether this could be negative mathematically

raynelfss replied:
See this [comment](https://github.com/Qiskit/qiskit/pull/14883#discussion_r2961175612).

alexanderivrii reviewed (commented):
Thanks Phalak and Ray,

This is a pretty straightforward PR in the sense that it faithfully ports Python code to Rust, making it easy to compare the two versions line by line.

I have left a few questions about some of the changes - Ray, can you please see if these are easy to answer/to address. If not, we can defer to the early 2.5 cycle.

Also, do you think we need release notes for this?

raynelfss committed:
Fix: Address review comments
- Added check for `Delay` and directive operations in `_push_node`.
- Use `saturating_sub` when calculating the shift variable in the same method.

raynelfss commented:
Thank you for taking the time to review this @alexanderivrii. I have addressed your comments and concerns to the best of my ability.

> Also, do you think we need release notes for this?

I don't believe it will be necessary here since, as you pointed out, this is a one-to-one port of the original pass and,  because nothing has changed on the client-facing side of things, we should not expose this in a release note.

alexanderivrii reviewed (approved):
Thanks Ray for the quick updates, LGTM!
